### PR TITLE
Improve include processing efficiency

### PR DESCRIPTION
**Pull request by jamesagnew**

### CODE REVIEW

**Comment by robogary** [comment]
Formatting check succeeded!

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java [hunk 1]
```diff
@@ -2209,20 +2218,66 @@ CanonicalUrlTargets calculateIndexUriIdentityHashesForResourceTypes(
...
+	 * can point to a Reference element (in which case we only need to _include actual
+	 * references by ID).
+	 * <p>
+	 * This isn't perfect because there's really no definitive and comprehensive
+	 * way of determining the datatype that a SearchParameter or a FHIRPath point to. But
+	 * we do our best if the path is simple enough to just manually check the type it
+	 * points to, or if it ends in an explicit type declaration.
+	 * </p>
+	 */
+	private boolean referencePathCouldPotentiallyReferenceCanonicalElement(
```
Comment (michaelabuckley) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java]: suggestion: can we move this to another class and give it some tests?  I don't know enough to review this well without some positive and negative examples.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java [hunk 2]
```diff
@@ -2209,20 +2218,66 @@ CanonicalUrlTargets calculateIndexUriIdentityHashesForResourceTypes(
 		return new CanonicalUrlTargets(hashIdentityValues, partitionIds);
 	}
 
-	static class CanonicalUrlTargets {
+	/**
+	 * Just because a SearchParameter is a Reference SP, doesn't necessarily mean that it
```
Comment (michaelabuckley) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java]: Nice docs!

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java [hunk 3]
```diff
@@ -2209,20 +2218,66 @@ CanonicalUrlTargets calculateIndexUriIdentityHashesForResourceTypes(
...
+	 * can reference a canonical. So first we try to rule out the SP based on the path it
+	 * contains. This matters because a SearchParameter of type Reference can point to
+	 * a canonical element (in which case we need to _include any canonical targets). Or it
+	 * can point to a Reference element (in which case we only need to _include actual
+	 * references by ID).
+	 * <p>
+	 * This isn't perfect because there's really no definitive and comprehensive
```
Comment (michaelabuckley) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java]: whine: this is such a drag.  I wish I could get the type pointed by a fhir-path expression elsewhere.  Is it impossible, or just hard?

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java [hunk 4]
```diff
@@ -1918,40 +1927,45 @@ private void loadIncludesMatchAll(
...
 
-				if (resourceId != null) {
-					JpaPid pid = JpaPid.fromIdAndVersionAndResourceType(resourceId, version, resourceType);
-					pid.setPartitionId(partitionId);
-					pidsToInclude.add(pid);
-				} else if (resourceCanonicalUrl != null) {
-					if (canonicalUrls == null) {
-						canonicalUrls = new HashSet<>();
+					Long version = null;
+					Long resourceId = (Long) ((Object[]) nextRow)[0];
```
Comment (michaelabuckley) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java]: suggestion: since your are here, consider replacing the generic Tuple for this row with a record type.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java [hunk 5]
```diff
@@ -2139,47 +2164,31 @@ CanonicalUrlTargets calculateIndexUriIdentityHashesForResourceTypes(
...
-							BaseRuntimeElementDefinition<?> childDef = child.getChildByName(elementName);
-							if (childDef != null) {
-								if (childDef.getName().equals("Reference")) {
-									continue;
-								}
-							}
+						.toList()) {
+
+					String paths = next.getPath();
+					for (String path : SearchParameterUtil.splitSearchParameterExpressions(paths)) {
```
Comment (michaelabuckley) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java:2194]: suggestion: this could be in a separate class without loss of readability.

**Review by michaelabuckley — APPROVED** [review]

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java (hunk 3 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java]: I guess both, depending on the specific expression. In some cases it's hard, in some cases impossible.

Examples:

* `Patient.generalPractitioner` - easy to know what the datatype will be
* `AuditEvent.entity.what` - hard-ish because the FHIRPath evaluator won't descend into a model unless it actually has content so we have to manually descend through the path
* `AuditEvent.entity.where(role=http://blah|123).what` - impossible because there is no way of knowing what that where predicate will actually end up picking in a real instance
* `AuditEvent.entity.what.as(Reference)` - hard-ish but possible because we can special case the tail 
* `Patient.extension('http://foo').value` -  impossible to know what the datatype will be

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java (hunk 1 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java]: Good call. Moved to SearchParameterUtil

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java (hunk 4 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java]: Done

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java (hunk 5 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/builder/SearchBuilder.java:2194]: Gave this a shot, but it uses several instance variables and just didn't seem possible to extract without making things less readable
